# Model Packaging und Containerisierung

In diesem Notebook wird gezeigt, wie das trainierte und registrierte Loan-Default-Modell in eine reproduzierbare Anwendungsumgebung überführt wird.

Der Schwerpunkt liegt auf drei Punkten:

- Laden des Modells aus der MLflow Model Registry
- Einbindung des Modells in die FastAPI-Anwendung
- Reproduzierbarer Start über Docker und Docker Compose

Damit bildet dieses Notebook die Brücke zwischen Modellentwicklung und Bereitstellung.


In [12]:
import mlflow
import pandas as pd
import numpy as np
from mlflow.tracking import MlflowClient
import os
from pathlib import Path
import sys

PROJECT_ROOT = Path("..").resolve()
SRC_DIR = PROJECT_ROOT / "src"
if str(SRC_DIR) not in sys.path:
    sys.path.append(str(SRC_DIR))


from loan_mlops.config import DEFAULT_MODEL_NAME, DEFAULT_TRACKING_URI
from loan_mlops.features import preprocess_api_payload

MODEL_NAME = DEFAULT_MODEL_NAME
TRACKING_URI = DEFAULT_TRACKING_URI

# MLflow Setup
mlflow.set_tracking_uri(TRACKING_URI)
client = MlflowClient()

# Model-Versionen laden
versions = client.search_model_versions(f"name='{MODEL_NAME}'")
for v in versions:
    print(
        "Version:", v.version,
        "| Stage:", v.current_stage,
        "| Run ID:", v.run_id,
        "| Source:", v.source
    )


Version: 1 | Stage: Staging | Run ID: ae0891ca882945f5b33c63525a0d5716 | Source: models:/m-32709742c9ae4dd7ac04e85ed2a83660


## 1. Modell laden und überprüfen



In [13]:
staging_versions = [v for v in versions if v.current_stage == "Staging"]

if not staging_versions:
    raise RuntimeError("Kein Modell in Stage 'Staging' gefunden.")

selected_version = sorted(staging_versions, key=lambda v: int(v.version), reverse=True)[0]

model = mlflow.sklearn.load_model(
    model_uri=f"models:/{MODEL_NAME}/{selected_version.version}"
)

print("Geladene Modellversion:", selected_version.version)
print("Run ID:", selected_version.run_id)
print(model)

Geladene Modellversion: 1
Run ID: ae0891ca882945f5b33c63525a0d5716
Pipeline(steps=[('scaler', StandardScaler()),
                ('model',
                 LogisticRegression(class_weight='balanced', max_iter=1000,
                                    random_state=42))])


## 2. Inferenz mit einem Beispielkredit prüfen

Bevor das Modell in der API verwendet wird, wird eine einzelne Testvorhersage ausgeführt.

Dabei wird dieselbe Vorverarbeitungslogik verwendet, die später auch in der API genutzt wird.  
Das ist wichtig, damit zwischen Notebook, API und späterer Anwendung keine Inkonsistenzen entstehen.

In [14]:
# Beispieldaten

sample_payload = {
    "credit_policy": 1,
    "purpose": "debt_consolidation",
    "int_rate": 0.12,
    "installment": 250.0,
    "log_annual_inc": 10.5,
    "dti": 14.0,
    "fico": 720,
    "days_with_cr_line": 4200.0,
    "revol_bal": 12000.0,
    "revol_util": 35.0,
    "inq_last_6mths": 1,
    "delinq_2yrs": 0,
    "pub_rec": 0,
}

input_df = preprocess_api_payload(sample_payload)
input_df.head()

,credit.policy,int.rate,installment,log.annual.inc,dti,fico,days.with.cr.line,revol.bal,revol.util,inq.last.6mths,...,debt_burden_ratio,dti_to_fico,int_rate_to_fico,revol_bal_to_inc,purpose_credit_card,purpose_debt_consolidation,purpose_educational,purpose_home_improvement,purpose_major_purchase,purpose_small_business
0,1,0.12,250.0,10.5,14.0,720,4200.0,12000.0,35.0,1,...,0.096378,0.019444,0.000167,0.330437,0,1,0,0,0,0


In [15]:
# Beispielvorhersage

prediction = model.predict(input_df)[0]

probability = None
if hasattr(model, "predict_proba"):
    probability = float(model.predict_proba(input_df)[0][1])

print("Vorhersage:", int(prediction))
print("Ausfallwahrscheinlichkeit:", probability)

Vorhersage: 0
Ausfallwahrscheinlichkeit: 0.40330099798884245


## 3. Zusammenhang zwischen Modell und API

Die eigentliche Bereitstellung erfolgt nicht im Notebook selbst, sondern über die FastAPI-Anwendung im Projektcode.

Die wichtigsten Dateien dafür sind:

- `src/loan_mlops/api.py`
- `main.py`

Die API übernimmt dabei folgende Aufgaben:

- registriertes Modell aus MLflow laden
- Eingabedaten validieren
- Payload in Modell-Features umwandeln
- Vorhersagen und Wahrscheinlichkeiten zurückgeben

Das Notebook dient hier vor allem dazu, die Architektur zu dokumentieren und die Zusammenhänge verständlich zu machen.

In [16]:
api_path = PROJECT_ROOT / "src" / "loan_mlops" / "api.py"
print(api_path.read_text()[:5000])

from __future__ import annotations

import logging
import time
from datetime import datetime
from types import SimpleNamespace
from typing import List, Literal

import pandas as pd
from fastapi import FastAPI, HTTPException, Request
from fastapi.middleware.cors import CORSMiddleware
from fastapi.responses import JSONResponse
from pydantic import BaseModel, Field

from .features import MODEL_COLUMNS, preprocess_api_payload
from .config import DEFAULT_MODEL_NAME, DEFAULT_TRACKING_URI

logger = logging.getLogger("loan_api")
logging.basicConfig(level=logging.INFO)


class LoanFeatures(BaseModel):
    credit_policy: int = Field(..., ge=0, le=1, description="1 = Kreditrichtlinie erfüllt")
    purpose: Literal[
        "all_other",
        "credit_card",
        "debt_consolidation",
        "educational",
        "home_improvement",
        "major_purchase",
        "small_business",
    ]
    int_rate: float = Field(..., ge=0, le=1, description="Zinssatz als Anteil, z. B. 0.12")
    install

In [17]:
main_path = PROJECT_ROOT / "main.py"
print(main_path.read_text())

from pathlib import Path
import sys

PROJECT_ROOT = Path(__file__).resolve().parent
SRC_DIR = PROJECT_ROOT / "src"
if str(SRC_DIR) not in sys.path:
    sys.path.append(str(SRC_DIR))

from loan_mlops.api import create_app

app = create_app()



## 4. Dockerfile

Für die containerisierte Ausführung des Projekts wird ein gemeinsames Dockerfile verwendet.

Dieses Dockerfile definiert die Laufzeitumgebung für die Anwendung, insbesondere:

- Python-Version und Systemabhängigkeiten
- Installation der Python-Packages aus `requirements.txt`
- Einbindung des Projektcodes

Das Dockerfile wird in diesem Projekt für mehrere Services genutzt:

- den MLflow Tracking Server
- die FastAPI-Anwendung

Beide Services verwenden dasselbe Basis-Image, werden jedoch mit unterschiedlichen Startbefehlen ausgeführt.

Der Vorteil dieses Ansatzes ist, dass alle Services auf einer konsistenten Umgebung basieren und keine Unterschiede zwischen Entwicklung und Ausführung entstehen.

In [18]:
dockerfile_path = PROJECT_ROOT / "Dockerfile"
print(dockerfile_path.read_text())

FROM python:3.11-slim

WORKDIR /app

ENV PYTHONDONTWRITEBYTECODE=1 \
    PYTHONUNBUFFERED=1 \
    PYTHONPATH=/app/src

RUN apt-get update && apt-get install -y --no-install-recommends \
    build-essential \
    gcc \
    && rm -rf /var/lib/apt/lists/*

COPY requirements.txt .
RUN pip install --no-cache-dir -r requirements.txt && \
    pip install --no-cache-dir jupyterlab notebook ipykernel

COPY . .



## 5. Docker Compose

Während das Dockerfile die Umgebung definiert, wird mit Docker Compose festgelegt, welche Services im Projekt gestartet werden.

In diesem Projekt werden drei zentrale Services verwendet:

- **mlflow**: Tracking-Server und Model Registry
- **api**: FastAPI-Anwendung zur Bereitstellung von Vorhersagen
- **streamlit**: Benutzeroberfläche für die interaktive Nutzung des Modells (siehe Notebook 8)

Beide Services werden aus demselben Dockerfile gebaut, unterscheiden sich jedoch in ihrem jeweiligen Startbefehl.

Der MLflow-Service wird mit einem MLflow-Server gestartet und stellt Tracking sowie Model Registry bereit.  
Die API wird mit Uvicorn gestartet und stellt Endpunkte für Vorhersagen zur Verfügung.
Die Streamlit-Anwendung greift auf die API zu und macht das System über eine grafische Oberfläche nutzbar.

Ein weiterer wichtiger Punkt ist die Kommunikation zwischen den Services:

- Innerhalb von Docker Compose wird der Service-Name (z. B. `mlflow`) als Hostname verwendet
- Außerhalb von Docker (z. B. im Notebook) erfolgt der Zugriff über `localhost` und den gemappten Port

Damit entsteht eine klare Trennung zwischen interner und externer Kommunikation.

In [19]:
compose_path = PROJECT_ROOT / "docker-compose.yml"
print(compose_path.read_text())

services:
  mlflow:
    build: .
    container_name: loan-mlops-mlflow
    working_dir: /app
    environment:
      PYTHONPATH: /app/src
    command:
      [
      "mlflow",
      "server",
      "--host", "0.0.0.0",
      "--port", "5000",
      "--backend-store-uri", "sqlite:////mlflow/mlflow.db",
      "--artifacts-destination", "/mlflow/artifacts",
      "--serve-artifacts"
      ]
    ports:
      - "5001:5000"
    volumes:
      - .:/app
      - mlflow_data:/mlflow
    healthcheck:
      test: ["CMD", "python", "-c", "import socket; s=socket.create_connection(('127.0.0.1', 5000), 2); s.close()"]
      interval: 5s
      timeout: 3s
      retries: 20
      start_period: 10s
    restart: unless-stopped

  api:
    build: .
    container_name: loan-mlops-api
    working_dir: /app
    depends_on:
      mlflow:
        condition: service_healthy
    environment:
      PYTHONPATH: /app/src
      MLFLOW_TRACKING_URI: http://mlflow:5000
      MLFLOW_EXPERIMENT_NAME: Loan Default Predicti

## 6. Ports und Service-Architektur

Die verschiedenen Komponenten des Systems laufen auf unterschiedlichen Ports:

- MLflow Tracking Server: intern Port 5000, extern z. B. 5001
- FastAPI-Anwendung: Port 8000
- Streamlit-Dashboard: Port 8501

Für den Zugriff gilt:

- **Im Notebook (lokal)**: Zugriff über `http://localhost:<port>`
- **Innerhalb von Docker Compose**: Zugriff über den Service-Namen, z. B. `http://mlflow:5000`

Diese Unterscheidung ist wichtig, da sich der Zugriffspfad je nach Ausführungsumgebung unterscheidet.

Die folgende Übersicht zeigt die Architektur:

- Notebook → MLflow über `localhost`
- API-Container → MLflow über `mlflow`
- Streamlit-Container → API über `api`
- Nutzer → Streamlit über `localhost:8501`

## 6. Typische Start- und Testbefehle

Typischer Ablauf:

1. MLflow, API, Streamlit starten (einzeln oder gemeinsam)
2. Health-Check ausführen
3. Vorhersage testen
4. Streamlit-Dashboard aufrufen (siehe Notebook 8)

Die folgenden Befehle zeigen die wichtigsten Schritte:

```bash
docker compose up -d mlflow
docker compose up -d api
docker compose up -d streamlit
docker compose up -d --build
docker compose ps
curl http://localhost:8000/health
http://localhost:8501